### ISSUE TITLE
Implement the CSS shape path editor for the images

### ISSUE THREAD

**Issue by oleq** [ckeditor/ckeditor5-image#217]
The [PoC](https://twitter.com/ckeditor/status/1012334778862002177) inspired by the [editor in Firefox](https://twitter.com/jensimmons/status/1006573790254338048) looks like something worth developing and stabilizing. The number of likes and RTs makes it the most awaited feature so far.

![spe-v2-medium](https://user-images.githubusercontent.com/1099479/42095878-29a2b172-7bb4-11e8-8d71-c7fc2c9694ae.gif)

## Things to consider

1. The PoC uses [`shape-outside`](https://developer.mozilla.org/en-US/docs/Web/CSS/shape-outside) CSS property. It interacts with the surroundings when the element is floated. Many shapes could be used (`circle`, `ellipse`, `polygon`) but each one of them requires a new UI. You can't edit the ellipse the same way you edit polygons.

   I'm for sticking to `polygon` which is the most flexible solution as one can approximate a circle or whatever regular shape they want with the right number of vertices.

2. [`clip-path`](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path) is another story. It can be used to use just a subset of an image but it does not interact with the text around. Supporting both modes (`shape-outside` and `clip-path`) could be an overkill and for sure it would require a complex UI/UX to switch between the modes. So `shape-outside` first, at least in the MVP.

3. We need some way to reset the polygon so it contains the entire image, i.e. the feature becomes disabled and not included in data. Some kind of a button (in the image toolbar) would help us a lot.

4. The PoC uses the arbitrary number of vertices. In reality, we need a possibility to create more or less than that. I research the possibilities and it seems quite simple (at least for the MVP). 
   * A new vertex indicator will show up when the mouse approaches one of the lines.
   ![group 4](https://user-images.githubusercontent.com/1099479/42095932-4f2f67d2-7bb4-11e8-9b02-10e8cf93900a.png)

   * Double-clicking an existing vertex will remove it from the polygon. Or, alternatively, clicking the vertex will select it and <kbd>⌫</kbd> will remove it. Or both.

5. Do we want this feature to be enabled as soon as the image is selected? Or maybe a toggle button in the toolbar should enable it? (could also reset the polygon)

6. The PoC uses [SVG.js](http://svgjs.com/elements/#svg-marker) for the polygon editor. It's an external library and we use but a fraction of its features. I'm not sure including it in the editor is the best solution (weight, maintenance in the future) especially considering that what I used in the PoC is fairly simple API we could develop as a part of UI library.

**Comment by scofalik** [comment]
I agree with almost everything you wrote, except for creating/removing vertices. For MVP, I'd simply stick with 8 vertices - 4 corners and 4 between them.

**Comment by oleq** [comment]
> I agree with almost everything you wrote, except for creating/removing vertices. For MVP, I'd simply stick with 8 vertices - 4 corners and 4 between them.

4 isn't enough: I checked with various image shapes and it does not make sense, way too limited. 8, as in the PoC, feels much better. Anyway, it's not like they cost us anything, it's 1LOC to add another vertex to the polygon.

**Comment by scofalik** [comment]
That's what I wrote. 8 vertices - 4 starting in corners and 4 between them.

>  Anyway, it's not like they cost us anything, it's 1LOC to add another vertex to the polygon.

Of course, I am quite sure that starting with 4, 8 or 20 vertices is not a problem. I just wouldn't play with creating UI for all of this in MVP. Unless it will not take a long time, of course.

**Comment by oleq** [comment]
> That's what I wrote. 8 vertices - 4 starting in corners and 4 between them.

My bad.

> Unless it will not take a long time, of course.

It won't. The initial number of nodes is just a decision, not an actual work. 

Adding and removing nodes is _some_ work but as I started coding it, it isn't much. 

A noticeable work would be if we went with other shapes (circles, rectangles) and decided to provide dedicated UI, e.g. to scale it, rotate, etc.

**Comment by scofalik** [comment]
> Adding and removing nodes is some work but as I started coding it, it isn't much.

But this is assuming that we stick with this SVG.js library, right?

**Comment by oleq** [comment]
> But this is assuming that we stick with this SVG.js library, right? Or is UI written by you?

SVG.js gives us API for primitives: lines, paths, circles, other geometric shapes + DOM event listeners. It does not provide a "polygon editor" out of the box. Some code is needed to bind stuff together into a polygon editor and that's what PoC does.

By writing

> Adding and removing nodes is some work but as I started coding it, it isn't much.

I meant that regardless of the base library, it's not a rocket science. 100LOC tops.
